### DISCUSSION TITLE
Allow to choose a different language for order documents

### DISCUSSION THREAD

**Discussion by CArlvJ1N0N8** [PrestaShop/PrestaShop#32239] [Ideas]
**Describe the bug**

Hello, 
I am using Prestashop in association with Amazon marketplace module. 
So Amazon orders are created inside of Prestashop by this module every time an order is available. 
Client is created at the same time. 
I have checked with the developer for this module and the client language is also imported. 
Although if i go to the client i can't see his language. 
My issue is i am connecting to the BO using french language in order to print the shipping notes for the Amazon orders. but all of them are printed in french. Even the emails sent to the customers are always in french. Although customers are US based or Canadian based (english talking).
I even changed the default language for my store to English. 
What do i need to do to be able to print or send emails to customers in their language.
How can i check a customer language? (maybe a feature to show customer language on customer page?)

---
**Edited on October 23rd by Louise B.**

A. ✅ If as a customer you log to the system and view your invoices/delivery slip it definitely needs to be in the customers language.
--> this should use the language used by the customer in FO (works OK already)

**Specifications:** https://github.com/PrestaShop/prestashop-specs/pull/161/files

B. ➡️ But if the dispatcher is preparing the shipping and printing the invoice/delivery slip, then does it need to be in his language or in the customer language since it will be put in the colis to be shipped.
--> we could provide a new option to let the merchant decide if he wants to use customer language or employee language to download delivery slips and invoices when preparing orders. This option is to be added in Shop parameters > Order settings with the following wording:

**Label**: `Default language for order documents` in `Admin.Shopparameters.Feature`
**Tooltip**: `Select in what language you want to download invoices and delivery slips.` in `Admin.Shopparameters.Help`
**Dropdown selector**: `Customer language` / `Logistician language` in `Admin.Shopparameters.Feature`

By default, the customer language is selected.

**Specifications:** https://github.com/PrestaShop/prestashop-specs/pull/152/commits/464be4a0ee88da1bbc3659e487d8640ac0726ee7

C. ✅ Then if it is an accountancy person who is printing the invoices/delivery slips i can understand it is printed in the accountant's language.
--> this should use the language used by the employee in BO (works OK already)

**Specifications:** https://github.com/PrestaShop/prestashop-specs/pull/161/files

**Additionnal information**

PrestaShop version: 1.7.4.2
PHP version: 7.0.27-0+deb9u1

**Comment by CArlvJ1N0N8** [comment]
Sorry a little error in what i said... 
On the customer page i can see the language for the customer is set to English, but when i print the required documents it prints in my language instead of the customer language. 

I tested by switching my language to English and it indeed printed the documents in English. 

But should it not print the documents in the customer language?

**Comment by khouloudbelguith** [comment]
Hi @CArlvJ1N0N8,

I manage to reproduce the issue with the last version 1.7.4.2.
It is an old issue, documents are printed in the employee language.
@marionf, this issue could be fixed in this issue: #9817?

Best regards, Khouloud

**Comment by marionf** [comment]
@colinegin 
What do you think about this suggestion ? Invoices and delivery slips should be printed in the customer's language instead of the employee's language ?

**Comment by CArlvJ1N0N8** [comment]
Hello, 

Any update for this ? 

Thanks,

**Comment by khouloudbelguith** [comment]
Hi @CArlvJ1N0N8,

Sorry not yet.
This issue needs to be specified by our developers.

Best regards, Khouloud

**Comment by colinegin** [comment]
Hi @CArlvJ1N0N8, please excuse the delay. 

I agree with you, the customer should be able to get his invoices / delivery slips in his own language when he needs to download them. However, those are also used by the employee (especially invoices, as  he needs to keep all of them for legal purposes) and I think he should have them in his own language too. 

@eternoendless / @marionf  is it possible to distinguish from where the invoice is downloaded and show it in the person's language (customer / merchant) ? Let me know if my question is unclear.

We will keep you updated @CArlvJ1N0N8 !

Have a nice day !

**Comment by CArlvJ1N0N8** [comment]
Let me ask the question differently. 

There are different ways to figure the situation. 
A. If as a customer you log to the system and view your invoices/delivery slip it definitely needs to be in the customers language. 

B. But if the dispatcher is preparing the shipping and printing the invoice/delivery slip, then does it need to be in his language or in the customer language since it will be put in the colis to be shipped. 

C. Then if it is an accountancy person who is printing the invoices/delivery slips i can understand it is printed in the accountant's language. 

For the situation A. the customer is connecting to his portion in the front office so he uses his own selected language but i need to validate if the documents are printed using his selected language. 

For the situation B. the dispatcher will print from the back-office for every order the invoice/delivery slip from the BO- Sell - Orders - orders screen and select the invoice and the delivery slip and this would need to be in the customer language, not the dispatcher language. But actually it is printed in the dispatchers language... 
![print invoice and delivery slip](https://user-images.githubusercontent.com/42921237/45702731-952f3480-bb40-11e8-92c9-0cc1fe723db7.JPG)

 For the situation C. the accountant will print all invoices and delivery slips using the print pdf function via BO- Sell - Orders - invoices screen and BO- Sell - Orders - delivery slips screen and print the selected according to his own language. 

Is this correct or is there another way to use the system and avoid any language issues?

**Comment by CArlvJ1N0N8** [comment]
Hello, 
Is there any update for this discussion?

**Comment by khouloudbelguith** [comment]
Hi @CArlvJ1N0N8,

This issue should be specified by our developers.
Thanks for understanding.

Best regards, Khouloud

**Comment by CArlvJ1N0N8** [comment]
Hello,
Is there any update for this discussion?
Did you get any return from the developers?

**Comment by colinegin** [comment]
Hello @CArlvJ1N0N8,

Sorry for the delay, we are currently quite busy with the upcoming 1.7.5 release !
Regarding the emails sent to customers, I have checked and they are always sent in the customer language. Do you know how emails are sent for Amazon ? Is it a module you use ? If yes, I think you should check with the module.

Regarding your point A and point C, it's already working this way (in the customer's language in front office, and in the used account's language in back-office).

For the point B, if I understand correctly your idea, you would like to be able to choose in the back-office the language in which we decide to print the invoice / delivery slip ? If this is correct, we need to think about it, as it might be complicated in terms of design. Furthermore, it would mean you need to check what is the language of your customer everytime you want to print an invoice.

What do you think ?

**Comment by CArlvJ1N0N8** [comment]
Hello, 

I am dealing with Amazon through the module Amazon Marketplace and it was already settled that they are sending emails in the customer defined language. 
Although they don't send the emails, it is the order status change that creates and sends the emails 

So it is not related to the Amazon module (except that they create the customer and provide the customer language) 

For this actual discussion the point is indeed about the account connected to the BO that is printing the documents to add them to the colis to be shipped. In this case the documents should be printed in the customer language.

**Comment by CArlvJ1N0N8** [comment]
Hello, 

Do we have any progress in this discussion? 

I am still wondering if i would do the corrections on my side and walk away from the standard prestashop version... 

Thanks, 

Carl

**Comment by colinegin** [comment]
Hello @CArlvJ1N0N8,

Thank you for your patience.

I've been discussing with our UX team, and we have came up with an idea. 
We could add in Shop parameters > Order settings the ability to choose between customer language and employee language to download delivery slips and invoices when preparing orders.

What do you think about it ?

Have a nice day,

**Comment by CArlvJ1N0N8** [comment]
Hello,


This seems perfect to me! 



Carl Valcke



4955 Rue de la Randonnée
J1N 0N8 Sherbrooke
QC, Canada
Cellulaire : + 1 819 239 89 40
Domicile : + 1 819 573 23 02
E-mail : carl@valcke.net

Envoyé de mon iPhone

**Comment by holle75** [comment]
.... might be smart to use DEFINED languages. NOT language of logged in BO user. Neither automatically customer language.
For example one of my shops is based in italy but Admin is using english BO. No one wants to change language of BO to italian to print invoices/documents or send attached invoices in "payment accepted"-eMail in italian? At least in this shop it needs to happen this way.

Some countries taxoffices still insist on getting invoices in local language (It seems this is against EU law but bureaucrats in those countries just gives a f.... ), so it should be up to the merchant in which languages he send and print and store invoices. It´s not a minor thing. Worst case your invoices won´t get accepted. I don´t know what the law says about the same invoice in different languages? It´s an official document. 

So a smart solution would be to give the merchant the option to choose languages either for customer and BO invoices/documents (besides language of logged in Admin). Any automatisation might be wrong.

1) automatically for BO and customer (but consistent ... also in attached invoice to "pament accepted" -eMail) ... for the lazy ones. Might be problematic. If bug/missing logic in attached invoice to "payment accepted"- eMail is solved it´s close to what it is now. BUT this makes it impossible to use a different language for Admin than local.

2) predefined languages for BO and customer
      a) language for BO (this should be independent from logged in Admin)
      b) language for customer (but consistent ... also in attached invoice to "pament accepted" -eMail)
                    2ba) either customers language or
                    2bb) predefined language 

Number 2) would be perfect imho 
what do you guys think?

**Comment by holle75** [comment]
PS. a fast workaround for this << bug/missing logic in attached invoice to "payment accepted"- eMail >> would be amazing.

**Comment by ShaimyWinda** [comment]
Hi @khouloudbelguith, 

I have the same issus. Were you able to correct this problem?

Thanks !

**Comment by khouloudbelguith** [comment]
Hi @ShaimyWinda,

Sorry, this issue is not fixed yet.
It is added to our bug roadmap.
PrestaShop is an open-source project, so it can be solved before if someone submits a pull request to solve it.

Thanks!

**Comment by clotairer** [comment]
@marionf Do you have some inputs about interfaces to choose the printing language ?
To help you we need wording and the look and feel.

Thank you.

**Comment by khouloudbelguith** [comment]
Partially fixed here: https://github.com/PrestaShop/PrestaShop/pull/18352

Thanks!

**Comment by Progi1984** [comment]
Fixed only for PDF sent by mail by #18352

**Comment by 20bekharweb** [comment]
Hi
I am using Prestashop 1.7 and In print order I have an awful design.
please help me to have and invoice order like something that I see in this page. Our language is Persian (RTL). please let me have a file to upload in PDF part of my server to solve this
![Capture](https://user-images.githubusercontent.com/66832909/84506343-79849e00-acd4-11ea-8d7d-bbc1b7aa03ee.JPG)

**Comment by colinegin** [comment]
@20bekharweb your problem is really different than the initial issue, could you please create a new issue so that we can investigate it ?

**Comment by colinegin** [comment]
Hello @holle75 @clotaire202 @CArlvJ1N0N8 , 

If I sum up the idea of this improvement : 

A. If as a customer you log to the system and view your invoices/delivery slip it definitely needs to be in the customers language.
--> this should use the language used by the customer in FO (works OK already)

B. But if the dispatcher is preparing the shipping and printing the invoice/delivery slip, then does it need to be in his language or in the customer language since it will be put in the colis to be shipped.
--> we could provide a new option to let the merchant decide if he wants to use customer language or employee language to download delivery slips and invoices when preparing orders. This option could be added in Shop parameters > Order settings, in a new block.

C. Then if it is an accountancy person who is printing the invoices/delivery slips i can understand it is printed in the accountant's language.
--> this should use the language used by the employee in BO (works OK already)

What do you think about it ?

**Comment by holle75** [comment]
https://github.com/PrestaShop/PrestaShop/issues/10258#issuecomment-466977515

**Comment by clotairer** [comment]
@colinegin it looks good for me.

@holle75 I understand that you need defined from a list of languages. Why not ?

So to uplaod in the backoffice, you could select invoice or delivrery slips language : 
- Customer language
- Employee language
- One of languages available on PrestaShop

**Comment by holle75** [comment]
>  you could select invoice or delivrery slips language :
> 
>     * Customer language
> 
>     * Employee language
> 
>     * One of languages available on PrestaShop

Thats the only option if you want to safely keep it legal overall in all countries.

**Comment by colinegin** [comment]
Why would some need another language than the customer or employee language ?

**Comment by CArlvJ1N0N8** [comment]
Hello,

I agree with the thinking as it is exactly what I was initially stating. 
But if you bring this situation to a usage situation, 
A. Is on the front and works already as it is using customer language.
B. When the dispatcher is using the delivery note to prepare the shipping he can print the delivery slips in 1. Bulk or 2. Per individual order. 
So you could say when individually printed it has to be in customer language instead of employee language as it will be placed in the shipping. 
The bulk solution is more discussable, it could be in the employee language, although I still think it is best to do it in the customer language as it can potentially be added to the shipping.

To my opinion these individual documents (printed in bulk or individually) are not meant for preparation but to be placed in the shipping. Certainly they can be used to facilitate the preparation but I am convinced there are some modules or other reports/exports that can do a better job so you can take all products in once and sort them by order, this would be done in the employee language. 
And it would be the same for the accountant language, to me no difference he would be connected as an employee.

But for the convenience, why not add a selector at the bulk printing to ask if employee or customer language? 

This is my 2 cents!

Carl Valcke



4955 Rue de la Randonnée
J1N 0N8 Sherbrooke
QC, Canada
Cellulaire : + 1 819 239 89 40
Domicile : + 1 819 573 23 02
E-mail : carl@valcke.net

**Comment by holle75** [comment]
> 
> 
> Why would some need another language than the customer or employee language ?

because not every admin in BO or every client in FO is using the language of the country the shop is situated in.
As i wrote it might happen (with high probabilty) that your invoices are not getting accepted by the tax office if not in "home country" language. As i wrote, this is against EU-Law but Reality. I don´t understand where´s the problem to just make, at least language of invoices (and why not everything) choosable (for merchant and for client) and not automatic? Where´s the advantage to predefine a language in regards of whatever person uses BO or FO? This should/could be a (legally complicated) option, not a must.

**Comment by LouiseBonnard** [comment]
> B. But if the dispatcher is preparing the shipping and printing the invoice/delivery slip, then does it need to be in his language or in the customer language since it will be put in the colis to be shipped.
--> we could provide a new option to let the merchant decide if he wants to use customer language or employee language to download delivery slips and invoices when preparing orders. This option could be added in Shop parameters > Order settings, in a new block.

Behavior is okay for me, here is what I suggest:

**Label**: `Default language for order documents` in `Admin.Shopparameters.Feature`
**Tooltip**: `Select in what language you want to download invoices and delivery slips.` in `Admin.Shopparameters.Help`
**Dropdown selector**: `Customer language` / `Logistician language` in `Admin.Shopparameters.Feature`

@Julievrz, does this wording look right for you? Once approved, I will add it to the specifications and add this issue to the backlog.

**Comment by Julievrz** [comment]
@LouiseBonnard It looks good to me ! 👍

**Comment by marionf** [comment]
@LouiseBonnard which option should be selected by default in the drop-down ?

**Comment by LouiseBonnard** [comment]
@marionf: the customer language.

**Comment by epatec** [comment]
Hello,
Is there any progress with this?
Should we wait for a solution or do it ourselves? Is there a way we could contribute?
Request: It should also be possible to set customer language during customer creation in BO.
Thanks,
Klaus

**Comment by marionf** [comment]
Hello @epatec 

PrestaShop is an open source solution, you are welcome if you want to develop this feature :)
You can look at the specifications here: https://github.com/PrestaShop/PrestaShop/issues/10258#issue-356511837
To contribute, you can make a pull request [here](https://github.com/PrestaShop/PrestaShop/tree/develop)

Thank you!

**Comment by cruftex** [comment]
Attention here:

PS is storing the product name in the order data itself. The name is copied from the product and depends on the language setting at the time of the creation. That means:

- If you print the invoice in a language other than it was originally created, it would have labels in the requested language but product text in the language of the order creation time.
- To reproduce the invoice that was actually sent out, you would need the language in use at that time, which could differ from the current customer and BO user language

The introduction of a switch would cause additional problems. The accountant would need to see the document in the company accounting language, while the customer support staff would like to see the original document sent to the customer.

Personally, I find it more important to be able to reproduce the invoice that was sent out to the customer.

What about:

- Add the language to the order
- Provide documents in the original language to the customer in FO
- Have two download options in the BO "Customer Invoice" and "Local Invoice"

Another thing to think about is maybe drop the feature that the product name is frozen in the order, or if that stays, consequently copy all language variants into the order product detail.
